modules/table/src/main/java/org/apache/ignite/internal/table/distributed/schema/CatalogValidationSchemasSource.java

```diff
@@ -0,0 +1,192 @@
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.table.distributed.schema;
+
+import static java.util.stream.Collectors.toList;
+
+import java.util.List;
+import java.util.Objects;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.function.Predicate;
+import java.util.stream.IntStream;
+import java.util.stream.Stream;
+import org.apache.ignite.internal.catalog.CatalogService;
+import org.apache.ignite.internal.catalog.descriptors.CatalogTableDescriptor;
+import org.apache.ignite.internal.hlc.HybridTimestamp;
+import org.apache.ignite.internal.schema.SchemaManager;
+
+/**
+ * An implementation over {@link CatalogService}.
+ */
+public class CatalogValidationSchemasSource implements ValidationSchemasSource {
+    private final CatalogService catalogService;
+
+    private final SchemaManager schemaManager;
+
+    private final ConcurrentMap<CatalogVersionsSpan, List<FullTableSchema>> catalogVersionSpansCache = new ConcurrentHashMap<>();
+
+    // TODO: Remove entries from cache when compacting Catalog https://issues.apache.org/jira/browse/IGNITE-20790
+    // TODO: Remove entries from cache when compacting schemas in SchemaManager https://issues.apache.org/jira/browse/IGNITE-20789
+    private final ConcurrentMap<CatalogVersionToTableVersionSpan, List<FullTableSchema>> catalogVersionToTableVersionSpansCache
+            = new ConcurrentHashMap<>();
+
+    /** Constructor. */
+    public CatalogValidationSchemasSource(CatalogService catalogService, SchemaManager schemaManager) {
+        this.catalogService = catalogService;
+        this.schemaManager = schemaManager;
+    }
+
+    @Override
+    public CompletableFuture<Void> waitForSchemaAvailability(int tableId, int schemaVersion) {
+        return schemaManager.schemaRegistry(tableId)
+                .schemaAsync(schemaVersion)
+                .thenApply(unused -> null);
+    }
+
+    @Override
+    public List<FullTableSchema> tableSchemaVersionsBetween(int tableId, HybridTimestamp fromIncluding, HybridTimestamp toIncluding) {
+        int fromCatalogVersion = catalogService.activeCatalogVersion(fromIncluding.longValue());
+        int toCatalogVersion = catalogService.activeCatalogVersion(toIncluding.longValue());
```

Comment:
Could you add a comment that it is safe for us to take catalog versions at a timestamps, since the external component should have already waited for schema synchronization.